docs/proposals/workflow.md

```diff
@@ -0,0 +1,281 @@
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+
+## Abstract
+
+A proposal to introduce [workflow](https://en.wikipedia.org/wiki/Workflow_management_system)
+functionality in kubernetes.
+Workflows (aka [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph) workflows
+since _tasks_ are organized in a Directed Acyclic Graph) are ubiquitous
+in modern [job schedulers](https://en.wikipedia.org/wiki/Job_scheduler), see for example:
+
+* [luigi](https://github.com/spotify/luigi)
+* [ozie](http://oozie.apache.org/)
+* [azkaban](https://azkaban.github.io/)
+
+Most of the [job schedulers](https://en.wikipedia.org/wiki/List_of_job_scheduler_software) offer
+workflow functionality to some extent.
+
+
+## Use Cases
+
+* As a user, I want to create a _JobB_ which depends upon _JobA_ running to completion.
+* As a user, I want workflow composability. I want to create a _JobA_ which will be triggered
+as soon as an already running workflow runs to completion.
+* As a user, I want to delete a workflow (eventually cascading to running _tasks_).
+* As a user, I want to debug a workflow (ability to track failure): in case a _task_
+didn't run user should have a way to backtrack the reason of the failure, understanding which
+dependency has not been satisified.
+
+
+## Implementation
+
+This proposal introduces a new REST resource `Workflow`. A `Workflow` is represented as a
+[graph](https://en.wikipedia.org/wiki/Graph_(mathematics)), more specifically as a DAG.
+Vertices of the graph represent steps of the workflow. The workflow steps are represented via a
+`WorkflowStep`<sup>1</sup> resource.
+The edges of the graph represent _dependecies_. To represent edges there is no explicit resource
+- rather they are stored as predecessors in each `WorkflowStep` (i.e. each node).
+The basic idea of this proposal consists in creation of each step postponing execution
+until all predecessors' steps run to completion.
+
+
+### Postponing execution
+
+At the time of writing, to defer execution there are two discussions in the community:
+[#17305](https://github.com/kubernetes/kubernetes/pull/17305): an
+_initializer_ is a dynamically registered object which permits to select a custom controller
+to be applied to a resource. The controller verifies the dependencies.
```

Comment:
I cannot tell if you are just pointing out the similarity to #17305 or if you are saying that Workflow will be implemented entirely via #17305, or that it will partly use #17305.

The WorkflowSpec has a nested JobSpec.  Creating a nested object is not the same as creating a bare object.  I don't see anything in #17305 about handling nested objects.  Therefore, at a minimum, I think that there would need to be a Workflow Controller that extracts the nested objects and creates them as top-level objects.  You could either have the Workflow Controller create them all immediately, and then use #17305 to sequence their start.  Or you could just use the Workflow Controller to create each nested object (step) only when it is ready.  I had assumed that you would do the latter, but rereading this section I'm not not sure what you intend to do.